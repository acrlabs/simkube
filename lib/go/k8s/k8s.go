package k8s

import (
	"fmt"
	"strings"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
)

func NewClient() (*kubernetes.Clientset, error) {
	config, err := rest.InClusterConfig()
	if err != nil {
		return nil, fmt.Errorf("could not get client config: %w", err)
	}

	k8sClient, err := kubernetes.NewForConfig(config)
	if err != nil {
		return nil, fmt.Errorf("could not initialize Kubernetes client: %w", err)
	}

	return k8sClient, nil
}

func NamespacedNameFromObjectMeta(objmeta metav1.ObjectMeta) string {
	return NamespacedName(objmeta.Namespace, objmeta.Name)
}

func NamespacedName(namespace, name string) string {
	return fmt.Sprintf("%s/%s", namespace, name)
}

func ProviderID(nodeName string) string {
	return fmt.Sprintf("simkube://%s", nodeName)
}

func SplitNamespacedName(namespacedName string) (string, string) {
	split := strings.SplitN(namespacedName, "/", 2)
	if len(split) == 2 {
		return split[0], split[1]
	} else {
		return split[0], ""
	}
}
