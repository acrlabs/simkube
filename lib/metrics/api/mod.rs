pub mod prometheus;

pub use prometheus::{
    Prometheus,
    PrometheusSpec,
    PrometheusStatus,
};
