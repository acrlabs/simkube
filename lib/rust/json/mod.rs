mod patch_ext;

pub use patch_ext::{
    patch_ext_add,
    patch_ext_remove,
};

#[cfg(test)]
mod patch_ext_test;
