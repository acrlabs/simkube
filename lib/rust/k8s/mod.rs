mod apiset;
mod container;
mod gvk;
pub mod macros;
mod pod;
mod pod_lifecycle;
mod util;

pub use apiset::*;
pub use gvk::*;
use k8s_openapi::api::core::v1 as corev1;
use k8s_openapi::apimachinery::pkg::apis::meta::v1 as metav1;
pub use util::*;

use crate::errors::*;
use crate::macros::partial_ord_eq_ref;

const LAST_APPLIED_CONFIG_LABEL_KEY: &str = "kubectl.kubernetes.io/last-applied-configuration";
const DEPL_REVISION_LABEL_KEY: &str = "deployment.kubernetes.io/revision";

err_impl! {KubernetesError,
    #[error("field not found in struct: {0}")]
    FieldNotFound(String),

    #[error("malformed container status: {0:?}")]
    MalformedContainerState(corev1::ContainerState),

    #[error("malformed label selector: {0:?}")]
    MalformedLabelSelector(metav1::LabelSelectorRequirement),
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum PodLifecycleData {
    Empty,
    Running(i64),
    Finished(i64, i64),
}
partial_ord_eq_ref!(PodLifecycleData);

pub trait KubeResourceExt {
    fn namespaced_name(&self) -> String;
    fn matches(&self, sel: &metav1::LabelSelector) -> anyhow::Result<bool>;
}

pub trait PodExt {
    fn spec(&self) -> anyhow::Result<&corev1::PodSpec>;
    fn status(&self) -> anyhow::Result<&corev1::PodStatus>;
    fn spec_mut(&mut self) -> &mut corev1::PodSpec;
    fn status_mut(&mut self) -> &mut corev1::PodStatus;
}

trait StartEndTimeable {
    fn start_ts(&self) -> anyhow::Result<Option<i64>>;
    fn end_ts(&self) -> anyhow::Result<Option<i64>>;
}

#[cfg(test)]
mod test;
