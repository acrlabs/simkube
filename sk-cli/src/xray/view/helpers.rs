use chrono::TimeDelta;
use lazy_static::lazy_static;
use ratatui::prelude::*;
use sk_core::prelude::*;

use crate::validation::{
    AnnotatedTraceEvent,
    ValidatorType,
};

pub(super) const LIST_PADDING: usize = 3;
lazy_static! {
    static ref ERR_STYLE: Style = Style::new().white().on_red();
    static ref WARN_STYLE: Style = Style::new().white().on_yellow();
}

pub(super) fn make_event_spans(event: &AnnotatedTraceEvent, start_ts: i64) -> (Span, Span) {
    let d = TimeDelta::new(event.data.ts - start_ts, 0).unwrap();
    let evt_span = Span::from(format!(
        "{} ({} applied/{} deleted)",
        format_duration(d),
        event.data.applied_objs.len(),
        event.data.deleted_objs.len()
    ));

    let (warnings, errs) = event.annotations.iter().fold((0, 0), |(mut w, mut e), codes| {
        for code in codes {
            match code.0 {
                ValidatorType::Warning => w += 1,
                ValidatorType::Error => e += 1,
            }
        }
        (w, e)
    });

    if warnings + errs == 0 {
        (evt_span, Span::default())
    } else {
        let err_str = format!("{} error{}", errs, if errs == 1 { "" } else { "s" });
        let warn_str = format!("{} warning{}", warnings, if warnings == 1 { "" } else { "s" });
        let style = if errs == 0 { *WARN_STYLE } else { *ERR_STYLE };
        let err_span = Span::styled(format!(" {err_str}/{warn_str} "), style);

        (evt_span, err_span)
    }
}

pub(super) fn make_object_spans<'a>(
    index: usize,
    obj: &DynamicObject,
    op: char,
    event: &'a AnnotatedTraceEvent,
) -> (Span<'a>, Span<'a>) {
    let obj_span = Span::styled(format!("  {} {}", op, obj.namespaced_name()), Style::new().italic());
    let (warnings, errs) = event.annotations[index].iter().fold((0, 0), |(mut w, mut e), code| {
        match code.0 {
            ValidatorType::Warning => w += 1,
            ValidatorType::Error => e += 1,
        }
        (w, e)
    });
    if warnings + errs == 0 {
        (obj_span, Span::default())
    } else {
        let err_str = format!("{} error{}", errs, if errs == 1 { "" } else { "s" });
        let warn_str = format!("{} warning{}", warnings, if warnings == 1 { "" } else { "s" });
        let style = if errs == 0 { *WARN_STYLE } else { *ERR_STYLE };
        let err_span = Span::styled(format!(" {err_str}/{warn_str} "), style);

        (obj_span, err_span)
    }
}

pub(super) fn format_list_entries<'a>(
    spans: impl Iterator<Item = (Span<'a>, Span<'a>)> + Clone,
    width: usize,
) -> Vec<Text<'a>> {
    let max_err_width = spans
        .clone()
        .max_by_key(|(_, err_span)| err_span.width())
        .map_or(0, |(_, err_span)| err_span.width());

    spans
        .map(|(mut item_span, err_span)| {
            // Instead of using a table we're just computing everything by hand here.  The format
            // of a list entry is LIST_PADDING|item|spaces|error|LIST_PADDING
            //
            // In order to ensure that all the error lines are the same width, we compute the
            // maximum above, and then use that everywhere.
            //
            // If the length of the item + the error + 2 * LIST_PADDING > width of the screen,
            // we have no spaces in the middle.  That's what these saturating subs are computing.
            // On the other hand, if there's no error to display, then we just fill out the rest of
            // the line with spaces.
            let mid_padding_width = width
                .saturating_sub(item_span.width())
                .saturating_sub(max_err_width)
                .saturating_sub(LIST_PADDING * 2);
            let mid_padding_span = Span::from(" ".repeat(mid_padding_width));

            // We want the right padding to match the same style as the error or warning, so if an
            // error is present we create a separate span here.
            let (trunc_width, right_padding_width) = match err_span.width() {
                0 => (width - LIST_PADDING, 0),
                x => (
                    width - (mid_padding_width + max_err_width + max_err_width - x + LIST_PADDING + LIST_PADDING),
                    max_err_width - x + LIST_PADDING,
                ),
            };
            let right_padding_span = Span::styled(" ".repeat(right_padding_width), err_span.style);

            // If the list entry is too long, then we truncate it and replace the rightmost
            // characters with an ellipsis.
            if trunc_width < item_span.content.len() {
                let content = item_span.content.to_mut();
                content.truncate(trunc_width);
                content.replace_range(trunc_width - 4..trunc_width, "... ");
            }

            Text::from(item_span.clone() + mid_padding_span + err_span.clone() + right_padding_span)
        })
        .collect()
}

pub(super) fn format_duration(d: TimeDelta) -> String {
    let day_str = match d.num_days() {
        x if x > 0 => format!("{x}d "),
        _ => String::new(),
    };

    format!("{}{:02}:{:02}:{:02}", day_str, d.num_hours() % 24, d.num_minutes() % 60, d.num_seconds() % 60)
}
