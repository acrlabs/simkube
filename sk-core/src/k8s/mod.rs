mod apiset;
mod container_state;
mod gvk;
mod lease;
mod owners;
mod pod_ext;
mod pod_lifecycle;
mod sim;
mod util;

pub use apiset::*;
pub use gvk::*;
pub use lease::*;
pub use owners::OwnersCache;
use serde::{
    Deserialize,
    Serialize,
};
pub use sim::*;
pub use util::*;

use crate::errors::*;
use crate::macros::partial_ord_eq_ref;
use crate::prelude::*;

const LAST_APPLIED_CONFIG_LABEL_KEY: &str = "kubectl.kubernetes.io/last-applied-configuration";
const DEPL_REVISION_LABEL_KEY: &str = "deployment.kubernetes.io/revision";

err_impl! {KubernetesError,
    #[error("field not found in struct: {0}")]
    FieldNotFound(String),

    #[error("lease has different owner: {0}")]
    LeaseHeldByOther(String),

    #[error("malformed container status: {0:?}")]
    MalformedContainerState(corev1::ContainerState),

    #[error("malformed label selector: {0:?}")]
    MalformedLabelSelector(metav1::LabelSelectorRequirement),
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub enum PodLifecycleData {
    Empty,
    Running(i64),
    Finished(i64, i64),
}
partial_ord_eq_ref!(PodLifecycleData);

pub trait KubeResourceExt {
    fn namespaced_name(&self) -> String;
    fn matches(&self, sel: &metav1::LabelSelector) -> anyhow::Result<bool>;
}

pub trait PodExt {
    fn spec(&self) -> anyhow::Result<&corev1::PodSpec>;
    fn stable_spec(&self) -> anyhow::Result<corev1::PodSpec>;
    fn status(&self) -> anyhow::Result<&corev1::PodStatus>;
}

trait StartEndTimeable {
    fn start_ts(&self) -> anyhow::Result<Option<i64>>;
    fn end_ts(&self) -> anyhow::Result<Option<i64>>;
}

#[cfg(test)]
pub mod tests;

#[cfg(feature = "testutils")]
pub mod testutils;
