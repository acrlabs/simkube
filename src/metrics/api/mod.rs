pub mod prometheus;
pub mod servicemonitor;

use std::collections::BTreeMap;

use kube::ResourceExt;
pub use prometheus::{
    Prometheus,
    PrometheusSpec,
    PrometheusStatus,
};
pub use servicemonitor::ServiceMonitor;
use servicemonitor::{
    ServiceMonitorEndpoints,
    ServiceMonitorEndpointsMetricRelabelings,
    ServiceMonitorEndpointsMetricRelabelingsAction,
    ServiceMonitorEndpointsRelabelings,
    ServiceMonitorEndpointsRelabelingsAction,
    ServiceMonitorEndpointsScheme,
    ServiceMonitorEndpointsTlsConfig,
    ServiceMonitorSpec,
};

use crate::k8s::build_object_meta;
use crate::prelude::*;

const PROM_VERSION: &str = "2.44.0";

pub fn build_ksm_service_monitor(name: &str, sim: &Simulation) -> anyhow::Result<ServiceMonitor> {
    // This object is just copy-pasta with minor modifications from the output of
    //
    // `kubectl describe servicemonitors kube-state-metrics`
    //
    // with the interval and scrape_timeout changed.  We may need to
    // adjust this more in the future.
    let mut metadata = build_object_meta(&sim.spec.monitoring_namespace, name, &sim.name_any(), sim)?;
    metadata
        .labels
        .get_or_insert(BTreeMap::new())
        .insert(APP_KUBERNETES_IO_NAME_KEY.into(), name.into());
    Ok(ServiceMonitor {
        metadata,
        spec: ServiceMonitorSpec {
            endpoints: vec![ServiceMonitorEndpoints {
                bearer_token_file: Some("/var/run/secrets/kubernetes.io/serviceaccount/token".into()),
                honor_labels: Some(true),
                interval: Some("1s".into()),
                metric_relabelings: Some(vec![ServiceMonitorEndpointsMetricRelabelings {
                    action: Some(ServiceMonitorEndpointsMetricRelabelingsAction::Drop),
                    regex: Some("kube_endpoint_address_not_ready|kube_endpoint_address_available".into()),
                    source_labels: Some(vec!["__name__".into()]),
                    ..Default::default()
                }]),
                port: Some("https-main".into()),
                relabelings: Some(vec![ServiceMonitorEndpointsRelabelings {
                    action: Some(ServiceMonitorEndpointsRelabelingsAction::LabelDrop),
                    regex: Some("(pod|service|endpoint|namespace)".into()),
                    ..Default::default()
                }]),
                scheme: Some(ServiceMonitorEndpointsScheme::Https),
                scrape_timeout: Some("1s".into()),
                tls_config: Some(ServiceMonitorEndpointsTlsConfig {
                    insecure_skip_verify: Some(true),
                    ..Default::default()
                }),
                ..Default::default()
            }],
            job_label: Some(APP_KUBERNETES_IO_NAME_KEY.into()),
            selector: metav1::LabelSelector {
                match_labels: Some(BTreeMap::from([(APP_KUBERNETES_IO_NAME_KEY.into(), "kube-state-metrics".into())])),
                ..Default::default()
            },
            ..Default::default()
        },
    })
}

pub fn build_prometheus(name: &str, svc_mon_selector: &str, sim: &Simulation) -> anyhow::Result<Prometheus> {
    Ok(Prometheus {
        metadata: build_object_meta(&sim.spec.monitoring_namespace, name, &sim.name_any(), sim)?,
        spec: PrometheusSpec {
            image: Some(format!("quay.io/prometheus/prometheus:v{}", PROM_VERSION)),
            service_monitor_selector: Some(metav1::LabelSelector {
                match_labels: Some(BTreeMap::from([(APP_KUBERNETES_IO_NAME_KEY.into(), svc_mon_selector.into())])),
                ..Default::default()
            }),
            service_account_name: Some(sim.spec.prometheus_service_account.clone()),
            version: Some(PROM_VERSION.into()),
            ..Default::default()
        },
        status: Default::default(),
    })
}
