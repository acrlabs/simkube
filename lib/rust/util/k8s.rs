use std::collections::BTreeMap;

use json_patch::{
    patch,
    PatchErrorKind,
    PatchOperation,
    RemoveOperation,
};
use k8s_openapi::apimachinery::pkg::apis::meta::v1 as metav1;
use kube::api::{
    DynamicObject,
    Resource,
    ResourceExt,
};
use tracing::*;

use crate::constants::SIMULATION_LABEL_KEY;
use crate::prelude::*;

pub fn add_common_fields<K>(sim_name: &str, owner: &K, obj: &mut impl Resource) -> SimKubeResult<()>
where
    K: Resource<DynamicType = ()>,
{
    obj.labels_mut().insert(SIMULATION_LABEL_KEY.into(), sim_name.into());
    obj.owner_references_mut().push(metav1::OwnerReference {
        api_version: K::api_version(&()).into(),
        kind: K::kind(&()).into(),
        name: owner.name_any(),
        uid: owner.uid().ok_or(SimKubeError::FieldNotFound)?,
        ..Default::default()
    });

    Ok(())
}

pub fn label_for(key: &str, val: &str) -> String {
    format!("{}={}", key, val)
}

pub fn namespaced_name(obj: &impl Resource) -> String {
    return match obj.namespace() {
        Some(ns) => format!("{}/{}", ns, obj.name_any()),
        None => obj.name_any().clone(),
    };
}

pub fn obj_matches_selector(obj: &impl Resource, sel: &metav1::LabelSelector) -> SimKubeResult<bool> {
    if let Some(exprs) = &sel.match_expressions {
        for expr in exprs {
            if !label_expr_match(obj.labels(), expr)? {
                return Ok(false);
            }
        }
    }

    if let Some(labels) = &sel.match_labels {
        for (k, v) in labels {
            if obj.labels().get(k) != Some(v) {
                return Ok(false);
            }
        }
    }
    Ok(true)
}

pub fn prefixed_ns(prefix: &str, obj: &impl Resource) -> String {
    format!("{}-{}", prefix, obj.namespace().unwrap())
}

pub fn strip_obj(obj: &mut DynamicObject, pod_spec_path: &str) -> SimKubeResult<()> {
    obj.metadata.uid = None;
    obj.metadata.resource_version = None;
    obj.metadata.managed_fields = None;
    obj.metadata.creation_timestamp = None;
    obj.metadata.deletion_timestamp = None;
    obj.metadata.owner_references = None;

    for suffix in &["nodeName", "serviceAccount", "serviceAccountName"] {
        let p = PatchOperation::Remove(RemoveOperation { path: format!("{}/{}", pod_spec_path, suffix) });
        if let Err(e) = patch(&mut obj.data, &[p]) {
            match e.kind {
                PatchErrorKind::InvalidPointer => {
                    debug!("could not find path {} for object {}, skipping", e.path, namespaced_name(obj));
                },
                _ => return Err(SimKubeError::JsonPatchError(e)),
            }
        }
    }

    Ok(())
}

pub fn split_namespaced_name(name: &str) -> (String, String) {
    match name.split_once('/') {
        Some((namespace, name)) => (namespace.into(), name.into()),
        None => ("".into(), name.into()),
    }
}

// The meanings of these operators is explained here:
// https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#set-based-requirement
pub(super) const OPERATOR_IN: &str = "In";
pub(super) const OPERATOR_NOT_IN: &str = "NotIn";
pub(super) const OPERATOR_EXISTS: &str = "Exists";
pub(super) const OPERATOR_DOES_NOT_EXIST: &str = "DoesNotExist";

pub(super) fn label_expr_match(
    obj_labels: &BTreeMap<String, String>,
    expr: &metav1::LabelSelectorRequirement,
) -> SimKubeResult<bool> {
    // LabelSelectorRequirement is considered invalid if the Operator is "In" or NotIn"
    // and there are no values; conversely for "Exists" and "DoesNotExist".
    return match expr.operator.as_str() {
        OPERATOR_IN => match obj_labels.get(&expr.key) {
            Some(v) => match &expr.values {
                Some(values) if !values.is_empty() => Ok(values.contains(v)),
                _ => Err(SimKubeError::MalformedLabelSelector),
            },
            None => Ok(false),
        },
        OPERATOR_NOT_IN => match obj_labels.get(&expr.key) {
            Some(v) => match &expr.values {
                Some(values) if !values.is_empty() => Ok(!values.contains(v)),
                _ => Err(SimKubeError::MalformedLabelSelector),
            },
            None => Ok(true),
        },
        OPERATOR_EXISTS => match &expr.values {
            Some(values) if !values.is_empty() => Err(SimKubeError::MalformedLabelSelector),
            _ => Ok(obj_labels.contains_key(&expr.key)),
        },
        OPERATOR_DOES_NOT_EXIST => match &expr.values {
            Some(values) if !values.is_empty() => Err(SimKubeError::MalformedLabelSelector),
            _ => Ok(!obj_labels.contains_key(&expr.key)),
        },
        _ => return Err(SimKubeError::MalformedLabelSelector),
    };
}
