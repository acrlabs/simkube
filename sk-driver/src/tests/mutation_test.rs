use std::collections::HashMap;

use json_patch_ext::{
    patch_ext,
    Patch,
};
use kube::api::TypeMeta;
use kube::core::admission::{
    AdmissionRequest,
    AdmissionResponse,
    AdmissionReview,
    Operation,
};
use kube::core::{
    GroupVersionKind,
    GroupVersionResource,
};
use mockall::predicate;
use rocket::serde::json::Json;
use sk_core::k8s::PodLifecycleData;
use sk_store::mock::MockTraceStore;

use super::helpers::build_driver_context;
use super::*;

#[fixture]
fn ctx(
    test_pod: corev1::Pod,
    #[default(vec![])] pod_owners: Vec<metav1::OwnerReference>,
    #[default(MockTraceStore::new())] store: MockTraceStore,
) -> DriverContext {
    let (_, client) = make_fake_apiserver();
    let mut owners = HashMap::new();
    owners.insert(test_pod.namespaced_name(), pod_owners);
    let cache = OwnersCache::new_from_parts(ApiSet::new(client), owners);
    build_driver_context(Arc::new(Mutex::new(cache)), Arc::new(store))
}

#[fixture]
fn adm_req(test_pod: corev1::Pod) -> AdmissionRequest<corev1::Pod> {
    let gvr = GroupVersionResource::gvr("".into(), "v1".into(), "pods".into());
    let gvk = GroupVersionKind::gvk("".into(), "v1".into(), "Pod".into());
    AdmissionRequest {
        types: TypeMeta { api_version: "v1".into(), kind: "Pod".into() },
        uid: "12345-12345".into(),
        kind: gvk,
        resource: gvr,
        sub_resource: None,
        request_kind: None,
        request_resource: None,
        request_sub_resource: None,
        name: test_pod.name_any(),
        namespace: Some(test_pod.namespace().unwrap()),
        operation: Operation::Create,
        user_info: Default::default(),
        object: Some(test_pod.clone()),
        old_object: None,
        dry_run: false,
        options: None,
    }
}

#[fixture]
fn adm_rev(adm_req: AdmissionRequest<corev1::Pod>) -> AdmissionReview<corev1::Pod> {
    AdmissionReview {
        types: Default::default(),
        request: Some(adm_req),
        response: None,
    }
}

#[fixture]
fn adm_resp(adm_req: AdmissionRequest<corev1::Pod>) -> AdmissionResponse {
    AdmissionResponse::from(&adm_req)
}

#[rstest]
#[traced_test]
#[tokio::test]
async fn test_handler_invalid_review(ctx: DriverContext) {
    let adm_rev = AdmissionReview {
        types: Default::default(),
        request: None,
        response: None,
    };
    let resp = handler(rocket::State::from(&ctx), Json(adm_rev), rocket::State::from(&MutationData::new())).await;
    assert!(!resp.0.response.unwrap().allowed);
}

#[rstest]
#[traced_test]
#[tokio::test]
async fn test_handler_bad_response(mut test_pod: corev1::Pod, mut adm_rev: AdmissionReview<corev1::Pod>) {
    let owner = metav1::OwnerReference {
        name: TEST_DRIVER_ROOT_NAME.into(),
        ..Default::default()
    };
    let ctx = ctx(test_pod.clone(), vec![owner.clone()], MockTraceStore::new());
    test_pod.owner_references_mut().push(owner);
    test_pod.spec = None;

    *adm_rev.request.as_mut().unwrap().object.as_mut().unwrap() = test_pod;
    let resp = handler(rocket::State::from(&ctx), Json(adm_rev), rocket::State::from(&MutationData::new())).await;
    assert!(!resp.0.response.unwrap().allowed);
}

#[rstest]
#[tokio::test]
async fn test_mutate_pod_not_owned_by_sim(mut test_pod: corev1::Pod, mut adm_resp: AdmissionResponse) {
    let owner = metav1::OwnerReference { name: "foo".into(), ..Default::default() };
    let ctx = ctx(test_pod.clone(), vec![owner.clone()], MockTraceStore::new());
    test_pod.owner_references_mut().push(owner);
    adm_resp = mutate_pod(&ctx, adm_resp, &test_pod, &MutationData::new()).await.unwrap();
    assert_eq!(adm_resp.patch, None);
}

#[rstest]
#[tokio::test]
async fn test_mutate_pod(mut test_pod: corev1::Pod, mut adm_resp: AdmissionResponse) {
    test_pod
        .annotations_mut()
        .insert(ORIG_NAMESPACE_ANNOTATION_KEY.into(), TEST_NAMESPACE.into());
    let root = metav1::OwnerReference {
        name: TEST_DRIVER_ROOT_NAME.into(),
        ..Default::default()
    };
    let depl = metav1::OwnerReference { name: TEST_DEPLOYMENT.into(), ..Default::default() };

    let owner_ns_name = format!("{TEST_NAMESPACE}/{TEST_DEPLOYMENT}");
    let mut store = MockTraceStore::new();
    let _ = store
        .expect_lookup_pod_lifecycle()
        .with(predicate::always(), predicate::always(), predicate::eq(EMPTY_POD_SPEC_HASH), predicate::eq(0))
        .returning(|_, _, _, _| PodLifecycleData::Finished(1, 2))
        .once();
    let _ = store.expect_has_obj().returning(move |_gvk, o| o == owner_ns_name);

    let ctx = ctx(test_pod.clone(), vec![root.clone(), depl.clone()], store);

    adm_resp = mutate_pod(&ctx, adm_resp, &test_pod, &MutationData::new()).await.unwrap();
    let mut json_pod = serde_json::to_value(&test_pod).unwrap();
    let pod_patch: Patch = serde_json::from_slice(&adm_resp.patch.unwrap()).unwrap();
    patch_ext(&mut json_pod, pod_patch.0[0].clone()).unwrap();
}
