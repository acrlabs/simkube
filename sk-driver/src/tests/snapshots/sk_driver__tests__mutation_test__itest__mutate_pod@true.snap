---
source: sk-driver/src/tests/mutation_test.rs
expression: json_pod
---
Object {
    "apiVersion": String("v1"),
    "kind": String("Pod"),
    "metadata": Object {
        "annotations": Object {
            "simkube.io/original-namespace": String("test-namespace"),
            "simkube.io/pod-sequence-number": String("0"),
            "simkube.io/pod-spec-stable-hash": String("17506812802394981455"),
            "simkube.kwok.io/stage-complete-time": String("1970-01-01T00:00:21+00:00"),
        },
        "labels": Object {
            "foo": String("bar"),
            "simkube.io/simulation": String("sk-test-driver-12345"),
            "simkube.kwok.io/stage-complete": String("true"),
        },
        "name": String("the-pod"),
        "namespace": String("test-namespace"),
    },
    "spec": Object {
        "containers": Array [],
        "nodeSelector": Object {
            "type": String("virtual"),
        },
        "tolerations": Array [
            Object {
                "effect": String("NoSchedule"),
                "key": String("kwok-provider"),
                "operator": String("Exists"),
            },
        ],
    },
    "status": Object {
        "phase": String("Running"),
    },
}
